File: 13-roman-to-integer/solution.py
#!/usr/bin/env python
# encoding: utf-8

class Solution:
    # @return an integer
    def romanToInt(self, s):
        symbol = {'I': 1,
                  'V': 5,
                  'X': 10,
                  'L': 50,
                  'C': 100,
                  'D': 500,
                  'M': 1000,
                 }

        total = 0
        prev = 'M'
        for num in s:
            if symbol[prev] < symbol[num]:
                total = total + symbol[num] - 2 * symbol[prev]
            else:
                total = total + symbol[num]
            prev = num

        return total

if __name__ == "__main__":
    str = "IV"
    s = Solution()
    print(s.romanToInt(str))

File: 13-roman-to-integer/solution.cpp
#include <iostream>
#include <string>

using namespace std;

class Solution {
public:
    int romanCharToInt(char c)
    {
        switch (c)
        {
            case 'I': return 1;
            case 'V': return 5;
            case 'X': return 10;
            case 'L': return 50;
            case 'C': return 100;
            case 'D': return 500;
            case 'M': return 1000;
        }
        return 0;
    }

    int romanToInt(string s) {

        int top = 0, sum = 0;
        int len = s.length();

        for (int i = len-1; i >= 0; i--)
        {
            int num = romanCharToInt(s[i]);
            if (num < top)
                sum -= num;
            else
                sum += num;
            top = num;
        }
        return sum;
    }
};

int main()
{
    string str = "MCMLXX";
    Solution s = Solution();
    cout << s.romanToInt(str) << endl;
}
